### PR TITLE
migrate foundation to nullsafety

### PR DESCRIPTION
**Pull request by a14n**
## Description

NNBD migration for foundation

## Related Issues

* dart-lang/sdk#40146

## Tests

I added the following tests:

None

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/foundation.dart [hunk 1]
```diff
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-// @dart = 2.8
+// @dart = 2.9
```
Comment (leafpetersen) [inline packages/flutter/lib/foundation.dart:5]: I think probably how this should be landed is by bumping the min sdk version in the pubspec to be a 2.9 release, and removing the // @dart comments in the opted in files entirely.  @jakemac53 @jonahwilliams does that sound right?

Comment (jonahwilliams) [reply packages/flutter/lib/foundation.dart:5]: That sounds like an approach with less churn. Is 2.9 _the_ null safety language version, or could that change?

Comment (leafpetersen) [reply packages/flutter/lib/foundation.dart:5]: 2.9 will *not* be the final null safety language version.  It is likely, but not guaranteed that 2.10 will be the final null safety language version.

Comment (jonahwilliams) [reply packages/flutter/lib/foundation.dart:5]: okay, in that case we should definitely do it in the pubspec so there is only a single location, right?

Comment (a14n) [reply packages/flutter/lib/foundation.dart:5]: I prefered to explicitly use `// @dart = 2.9` during the incremental migration for simple searchability. Once all framework is migrated I will remove it. Note that the [pubspec.yaml sdk version is already pointing to 2.9](https://github.com/flutter/flutter/blob/master/packages/flutter/pubspec.yaml#L8).

Does it look good to you?

Comment (jonahwilliams) [reply packages/flutter/lib/foundation.dart:5]: If it makes the process easier, I would say to use it - but file an issue to clean up the overrides once it has all landed.

Comment (leafpetersen) [reply packages/flutter/lib/foundation.dart:5]: I think that's better, and it's how pub packages need to migrate (because otherwise you're allowing sdks that don't support null safety, which won't work).  I think with flutter, because you provide your own copy of the sdk, you don't run into the versioning problem, but it still seems like the right approach to me.  But I don't know all the details of how package:flutter works - it's Not Like Other Packages.

Comment (leafpetersen) [reply packages/flutter/lib/foundation.dart:5]: Ok, works for me.  Note that if/when the Dart version bumps to a 2.10 pre-release, there will a bit of a bump - you won't be able roll 2.10 in until you change all of these from // @dart = 2.9 to // @dart = 2.10.

Comment (jonahwilliams) [reply packages/flutter/lib/foundation.dart:5]: could we set it to 2.10 now?

Comment (leafpetersen) [reply packages/flutter/lib/foundation.dart:5]: > could we set it to 2.10 now?

You could try it, but I wouldn't really recommend it.  You're more or less relying on undefined behavior in that case.  It might just work, in which case it's probably fine, but it also might work in some places and not others, or suddenly stop working, or whatever.

Comment (leafpetersen) [reply packages/flutter/lib/foundation.dart:5]: If this is really just about tracking which things are migrated, maybe consider just using some other arbitrary string to tag them?  // @dart = migrated ?

Comment (a14n) [reply packages/flutter/lib/foundation.dart:5]: Ok I'll remove it directly. I don't use it too much now.

File: packages/flutter/analysis_options.yaml [hunk 2]
```diff
@@ -5,3 +5,7 @@ include: ../analysis_options.yaml
 analyzer:
   enable-experiment:
     - non-nullable
+  errors:
+    always_require_non_null_named_parameters: false # not needed with nnbd
+    void_checks: false # issues with nnbd
```
Comment (goderbauer) [inline packages/flutter/analysis_options.yaml:10]: Is there an issue on file that we could link from here?

File: packages/flutter/analysis_options.yaml [hunk 3]
```diff
@@ -5,3 +5,7 @@ include: ../analysis_options.yaml
...
+    unnecessary_null_comparison: false
```
Comment (goderbauer) [inline packages/flutter/analysis_options.yaml:11]: Why do we turn this one off?

File: packages/flutter/lib/src/foundation/assertions.dart [hunk 4]
```diff
@@ -1005,17 +1004,17 @@ class FlutterError extends Error with DiagnosticableTreeMixin implements Asserti
       final String package = '${frame.packageScheme}:${frame.package}';
       if (removedPackagesAndClasses.containsKey(className)) {
         skipped += 1;
-        removedPackagesAndClasses[className] += 1;
+        removedPackagesAndClasses.update(className, (int value) => value + 1);
```
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/assertions.dart:1007]: Why is this change necessary?

File: packages/flutter/lib/src/foundation/object.dart [hunk 5]
```diff
@@ -2,19 +2,26 @@
...
-String objectRuntimeType(Object object, String optimizedValue) {
+String objectRuntimeType(Object? object, String optimizedValue) {
   assert(() {
     optimizedValue = object.runtimeType.toString();
     return true;
   }());
   return optimizedValue;
 }
+
+/// Allows promotion of a variable to another type.
```
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/object.dart:20]: This feels like a strange thing to add to flutter's public API. Looks like it is only used in one file. Can we just add it as a private thing there? Or, instead of using promote, can you just do the following in the code:

```
num n = 42;
n = n as int;
n.isEven;
```

File: packages/flutter/lib/src/foundation/observer_list.dart [hunk 6]
```diff
@@ -29,7 +27,7 @@ import 'dart:collection';
 class ObserverList<T> extends Iterable<T> {
   final List<T> _list = <T>[];
   bool _isDirty = false;
-  HashSet<T> _set;
+  late final HashSet<T> _set = HashSet<T>();
```
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/observer_list.dart:30]: Why `late`?

File: packages/flutter/lib/src/foundation/print.dart [hunk 7]
```diff
@@ -112,15 +110,16 @@ Iterable<String> debugWordWrap(String message, int width, { String wrapIndent =
...
+  final Match prefixMatch = _indentPattern.matchAsPrefix(message)!;
+  final String prefix = wrapIndent + ' ' * prefixMatch.group(0)!.length;
   int start = 0;
   int startForLengthCalculations = 0;
   bool addPrefix = false;
   int index = prefix.length;
   _WordWrapParseMode mode = _WordWrapParseMode.inSpace;
-  int lastWordStart;
-  int lastWordEnd;
+  // TODO(a14n): use `late int lastWordStart;`
```
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/print.dart:120]: What's currently stopping us to do this?

File: packages/flutter/lib/src/foundation/assertions.dart (hunk 4 above)
Comment (spkersten) [reply packages/flutter/lib/src/foundation/assertions.dart:1007]: `someMap[someKey]` is optional.

File: packages/flutter/analysis_options.yaml (hunk 2 above)
Comment (a14n) [reply packages/flutter/analysis_options.yaml:10]: link to issue added.

File: packages/flutter/analysis_options.yaml (hunk 3 above)
Comment (a14n) [reply packages/flutter/analysis_options.yaml:11]: This is needed to avoid warning on null assertion on non-nullable variable. For instance:

```dart
void m(String s) {
  assert(s != null);
}
```

This could change with https://github.com/flutter/flutter/issues/61042

File: packages/flutter/lib/src/foundation/assertions.dart (hunk 4 above)
Comment (a14n) [reply packages/flutter/lib/src/foundation/assertions.dart:1007]: Without this change there's an error because `removedPackagesAndClasses[className]` is nullable and you cannot use operator + on a nullable value. Without `.update` we should have used:

```dart
removedPackagesAndClasses[className] = removedPackagesAndClasses[className]! + 1;
```

The update version avoids repetitions and is less error prone imho.

File: packages/flutter/lib/src/foundation/object.dart (hunk 5 above)
Comment (a14n) [reply packages/flutter/lib/src/foundation/object.dart:20]: I don't remind the issue but @Hixie suggested to add this function.

File: packages/flutter/lib/src/foundation/observer_list.dart (hunk 6 above)
Comment (a14n) [reply packages/flutter/lib/src/foundation/observer_list.dart:30]: IIRC this will make the field lazy initialized.

File: packages/flutter/lib/src/foundation/print.dart (hunk 7 above)
Comment (a14n) [reply packages/flutter/lib/src/foundation/print.dart:120]: `late` is a little dangerous for public member. I'm not sure the TODO is still relevant. I'll check.

File: packages/flutter/lib/src/foundation/observer_list.dart (hunk 6 above)
Comment (jakemac53) [reply packages/flutter/lib/src/foundation/observer_list.dart:30]: It won't do anything in this case. It allows you to (yourself) lazily initialize it but leave it as a non-nullable type.

Comment (goderbauer) [reply packages/flutter/lib/src/foundation/observer_list.dart:30]: Sorry if this wasn't clear, but my question was meant as: "Why are we declaring this field as lazy? It seems unnecessary here."

File: packages/flutter/lib/src/foundation/assertions.dart (hunk 4 above)
Comment (goderbauer) [reply packages/flutter/lib/src/foundation/assertions.dart:1007]: Agreed, `update` does look nicer. Thanks for the explanation.

File: packages/flutter/analysis_options.yaml (hunk 3 above)
Comment (goderbauer) [reply packages/flutter/analysis_options.yaml:11]: If the variable itself is already non-nullable, why do we still need the assert? Wouldn't we want to clean that up and remove it?

File: packages/flutter/lib/src/foundation/observer_list.dart (hunk 6 above)
Comment (jakemac53) [reply packages/flutter/lib/src/foundation/observer_list.dart:30]: It was lazy before - I think there was a misunderstanding around what `late` means. If assigning immediately it has no meaning @stereotype441 should we have a lint for this?

To restore the lazy behavior that existed before this would have to be a nullable field type, and you would have to restore the old logic.

File: packages/flutter/lib/src/foundation/print.dart (hunk 7 above)
Comment (goderbauer) [reply packages/flutter/lib/src/foundation/print.dart:120]: I am not sure I fully understand, `lastWordStart` seems to be private to this method?

File: packages/flutter/analysis_options.yaml (hunk 3 above)
Comment (goderbauer) [reply packages/flutter/analysis_options.yaml:11]: Either way, It would probably be good to have a comment on this line explaining why we turn this off.

File: packages/flutter/lib/src/foundation/observer_list.dart (hunk 6 above)
Comment (leafpetersen) [reply packages/flutter/lib/src/foundation/observer_list.dart:30]: @jakemac53 that's not correct. late with an initializer means lazy.

Comment (goderbauer) [reply packages/flutter/lib/src/foundation/observer_list.dart:30]: Now I am confused. What is the difference between those two lines?

```dart
late final int foo = 1;
final int bar = 1;
```

Comment (leafpetersen) [reply packages/flutter/lib/src/foundation/observer_list.dart:30]: Nothing, because you can't observe laziness with a value.  But:

```dart
class A {
  late final int foo = (() {print("hello"); return 1;})();
  final int bar = (() {print("world"); return 1;})();
  A();
}
                            
                            
void main() {
  var a = A();
  print("says");
  a.foo;
}
```

outputs
```
world
says
hello
```

because a `late` variable with an initializer only evaluates its initializer on first use.

cc @munificent this is something we're going to have to explain carefully.

Comment (goderbauer) [reply packages/flutter/lib/src/foundation/observer_list.dart:30]: Thanks for the explanation!

File: packages/flutter/analysis_options.yaml (hunk 3 above)
Comment (a14n) [reply packages/flutter/analysis_options.yaml:11]: >If the variable itself is already non-nullable, why do we still need the assert?

See https://github.com/dart-lang/language/issues/1018:

> At first glance, it might seem sensible to remove the assertion on the grounds that c can no longer be null, but that's only true in strong mode. Flutter will have a long tail of users running in weak mode until Dart 3.0, and assertions like these are necessary to help prevent them from passing nulls to Flutter APIs.

File: packages/flutter/lib/src/foundation/object.dart (hunk 5 above)
Comment (a14n) [reply packages/flutter/lib/src/foundation/object.dart:20]: Finally I removed this function for now.

File: packages/flutter/lib/src/foundation/print.dart (hunk 7 above)
Comment (a14n) [reply packages/flutter/lib/src/foundation/print.dart:120]: comment removed. I added it to remind me that a lint could be done to use `late`.

**Comment by jakemac53** [comment]
Note that force pushing destroys all pr comment history so please avoid that :).

File: packages/flutter/analysis_options.yaml (hunk 3 above)
Comment (goderbauer) [reply packages/flutter/analysis_options.yaml]: Thanks for the context and adding it as a comment.

**Review by goderbauer — APPROVED** [review]
LGTM
